hazelcast-jet-core/src/main/java/com/hazelcast/jet/datamodel/DataModelSerializerHooks.java

```diff
@@ -0,0 +1,364 @@
+                }
+
+                @Override
+                public void destroy() {
+                }
+            };
+        }
+
+        @Override public boolean isOverwritable() {
+            return false;
+        }
+    }
+
+    public static final class TagHook implements SerializerHook<Tag> {
+
+        @Override
+        public Class<Tag> getSerializationType() {
+            return Tag.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new StreamSerializer<Tag>() {
+                @Override
+                public void write(ObjectDataOutput out, Tag tag) throws IOException {
+                    out.writeInt(tag.index());
+                }
+
+                @Override
+                public Tag read(ObjectDataInput in) throws IOException {
+                    return Tag.tag(in.readInt());
+                }
+
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.TAG;
+                }
+
+                @Override
+                public void destroy() {
+                }
+            };
+        }
+
+        @Override public boolean isOverwritable() {
+            return false;
+        }
+    }
+
+    public static final class ItemsByTagHook implements SerializerHook<ItemsByTag> {
+
+        @Override
+        public Class<ItemsByTag> getSerializationType() {
+            return ItemsByTag.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new ItemsByTagSerializer();
+        }
+
+        @Override public boolean isOverwritable() {
+            return false;
+        }
+
+    }
+
+    public static final class BagsByTagHook implements SerializerHook<BagsByTag> {
+
+        @Override
+        public Class<BagsByTag> getSerializationType() {
+            return BagsByTag.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new BagsByTagSerializer();
+        }
+
+        @Override public boolean isOverwritable() {
+            return false;
+        }
+
+    }
+
+    private static class ItemsByTagSerializer implements StreamSerializer<ItemsByTag> {
+        @Override
+        public void write(ObjectDataOutput out, ItemsByTag ibt) throws IOException {
+            Set<Entry<Tag<?>, Object>> entries = ibt.entrySet();
+            out.writeInt(entries.size());
+            for (Entry<Tag<?>, Object> e : entries) {
+                out.writeObject(e.getKey());
+                Object val = e.getValue();
+                out.writeObject(val != NONE ? val : null);
+            }
+        }
+
+        @Override
+        public ItemsByTag read(ObjectDataInput in) throws IOException {
+            int size = in.readInt();
+            ItemsByTag ibt = new ItemsByTag();
+            for (int i = 0; i < size; i++) {
+                ibt.put(in.readObject(), in.readObject());
```

Comment:
shouldn't it be set to `NONE` when reading?

Reply:
No because that's done internally by `put()` There's an asymmetry between writing and reading because writing is done by iterating the raw internal map, whereas reading uses the public API.